pennylane_forest/device.py

```diff
@@ -224,31 +224,53 @@ def program(self):
         """View the last evaluated Quil program"""
         return self.prog
 
-    def apply(self, operation, wires, par):
-        # pylint: disable=attribute-defined-outside-init
+    def remap_wires(self, wires):
+        """Use the wiring specified for the device if applicable.
+
+        Returns:
+            list: wires as integers corresponding to the wiring if applicable
+        """
         if hasattr(self, "wiring"):
-            qubits = [int(self.wiring[i]) for i in wires]
-        else:
-            qubits = [int(w) for w in wires]
+            return [int(self.wiring[i]) for i in wires]
+
+        return [int(w) for w in wires]
+
+    def apply(self, operations, **kwargs):
+        # pylint: disable=attribute-defined-outside-init
+        rotations = kwargs.get("rotations", [])
+
+        # Storing the active wires
+        self._active_wires = ForestDevice.active_wires(operations + rotations)
+
+        # Apply the circuit operations
+        for i, operation in enumerate(operations):
+            # number of wires on device
+            wires = self.remap_wires(operation.wires)
+            par = operation.parameters
+
+            if i > 0 and operation.name in ("QubitStateVector", "BasisState"):
+                raise DeviceError("Operation {} cannot be used after other Operations have already been applied "
+                                  "on a {} device.".format(operation.name, self.short_name))
+            self.prog += self._operation_map[operation.name](*par, *wires)
 
-        self.prog += self._operation_map[operation](*par, *qubits)
+        self.apply_rotations(rotations)
 
-        # keep track of the active wires. This is required, as the
-        # pyQuil wavefunction simulator creates qubits dynamically.
-        if wires:
-            self.active_wires = self.active_wires.union(set(wires))
-        else:
-            self.active_wires = set(range(self.num_wires))
+    def apply_rotations(self, rotations):
+        """Apply the circuit rotations.
 
-    @abc.abstractmethod
-    def pre_measure(self):  # pragma no cover
-        """Run the QVM or QPU"""
-        raise NotImplementedError
+        This method serves as an auxiliary method to :meth:`~.ForestDevice.apply`.
+        """
+        for operation in rotations:
+            wires = self.remap_wires(operation.wires)
+            par = operation.parameters
+            self.prog += self._operation_map[operation.name](*par, *wires)
 
     def reset(self):
         self.prog = Program()
-        self.active_wires = set()
-        self.state = None
+        self._active_wires = set()
+        self._state = None
+        self._parameter_map = {}
+        self._parameter_reference_map = {}
```

Comment:
What's the difference between the parameter map and parameter reference map?

Reply:
The parameter map stores the numeric value of the symbolic parameter, whereas the parameter reference map stores the PyQuil reference that has been created for it. I'll add docstrings explaining these differences